upcoming/composable-pipelines-pattern.md

```diff
@@ -0,0 +1,392 @@
+
+Still, for some problems it is nice to be able to rely solely on the standard-library, which lead the author to start experimenting with how far one can go with Flow-based programming inspired ideas without using any 3rd party framework at all - that is, finding out the most flow-based programming-like pattern one can implement in pure Go.
+
+By playing around, the first thing that became clear is that Go provides a lot more flexibility to how to define and wire together components than the with e.g. the generator functions in python.
+
+One of the patterns arising from this experimentation that the author tends to like a lot is one where concurrent processes are encapsulated into structs, and inputs and outputs defined are struct fields of type channel (of some subsequent type).
+
+This pattern lets us do all the wiring of the processes using the struct fields, which, by having clear names coded into the processes, and exposed to the network-defining code, can in the authors opinion make the network definition code clearer and more intuitive.
+
+This pattern is the subject for this post, and is presented in more detail below.
+
+
+### Show me the code
+
+So, what does this pattern look like in practice?
+
+Code examples of this little pattern can be found in a github repo that the author made for this idea, called [glow](https://github.com/samuell/glow), but let's have a look at the code examples here in the post as well:
+
+#### An example component
+
+First let's just have a look at how a component looks, when using this pattern. Every component has one or more "in-" and "out-ports", consisting of struct fields of type channel (of some subsequent type - `[]byte` arrays in this case).
+
+Then it has an `Init()` method that initializes a go-routine so that it is ready to start reading on the inport(s) and write on the outport(s) as soon as the first in-data arrive:
+
+````go
+package glow
+
+import (
+	"bufio"
+	"os"
+)
+
+type StdInReader struct {
+	Out chan []byte
+}
+
+func (self *StdInReader) Init() {
+	go func() {
+		scan := bufio.NewScanner(os.Stdin)
+		for scan.Scan() {
+			self.Out <- append([]byte(nil), scan.Bytes()...)
+		}
+		close(self.Out)
+	}()
+}
+````
+
+Here, since we did have just one output, (and no input), we have simply named the outport to "Out", but we could of course give it a more descriptive name if we wish.
+
+Notice also inside the go-routine (the `go func() {...}` bit) above, how even the internal component code becomes a bit clearer by using fields of the struct, so that we know immediately where to look up the fields definition!
+
+#### Connecting components - manual way
+
+To connect such processes together - here in the more "manual" way - we create a bunch of channels, a bunch of processes and finally stitch them together and run it:
+
+````go
+package main
+
+import (
+	"fmt"
+	"github.com/samuell/glow"
+)
+
+const (
+	BUFSIZE = 128 // Set a buffer size to use for channels
```

Comment:
My somewhat extensive testing has shown that performance is quite severely degraded with unbuffered channels. I don't have a clear number for the optimum, but 128 lies within the "not bad" range. I changed now to 16 in the code examples though, since this was a rather sharp limit, under which a clear trend for worse performance was seen.